source/dialogs/bulkconverterdialog.cpp

```diff
@@ -0,0 +1,236 @@
+  * the Free Software Foundation, either version 3 of the License, or
+  * (at your option) any later version.
+  *
+  * This program is distributed in the hope that it will be useful,
+  * but WITHOUT ANY WARRANTY; without even the implied warranty of
+  * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+  * GNU General Public License for more details.
+  *
+  * You should have received a copy of the GNU General Public License
+  * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+*/
+
+#include "bulkconverterdialog.h"
+#include "ui_bulkconverterdialog.h"
+
+#include "formats/powertab/common.h"
+
+#include <app/paths.h>
+#include <score/score.h>
+
+#include <QFileDialog>
+#include <QFileInfo>
+#include <QDebug>
+#include <QThread>
+
+std::optional<std::string> convertFile(const boost::filesystem::path& src,
+                                       const boost::filesystem::path& dst,
+                                       std::unique_ptr<FileFormatManager>& ffm)
+{
+    QFileInfo fileInfo(QString::fromStdString(src.string()));
+    std::optional<FileFormat> format = ffm->findFormat(
+                fileInfo.suffix().toStdString());
+
+    if (!format) return "bad format";
+
+    Score score;
+    try
+    {
+        ffm->importFile(score, src, *format);
+    }
+    catch (const std::exception &e)
+    {
+        return "could not import file: " + src.string();
+    }
+
+    try
+    {
+      ffm->exportFile(score, dst, FileFormat(getPowerTabFileFormat()));
+    }
+    catch (const std::exception &e)
+    {
+        return "could not export file: " + dst.string();
+    }
+
+    return std::nullopt;
+}
+
+BulkConverterWorker::BulkConverterWorker(boost::filesystem::path& source,
+                                         boost::filesystem::path& destination,
+                                         bool dryRun,
+                                         std::unique_ptr<FileFormatManager>& fileFormatManager)
+  : QObject(), mySrc(source), myDst(destination), myDryRun(dryRun),
+    myFileCount(0), myFileFormatManager(fileFormatManager)
+{
+}
+
+BulkConverterWorker::~BulkConverterWorker()
+{
+}
+
+void BulkConverterWorker::walkAndConvert()
+{
+    std::vector<boost::filesystem::path> children;
+    const std::string basePath = mySrc.string();
+
+    children.emplace_back(mySrc);
+
+    do {
+        boost::filesystem::path p = children.back();
+        children.pop_back();
+
+        if (!boost::filesystem::is_directory(p)) continue;
+
+        auto dir_it = boost::filesystem::directory_iterator(p);
+
+        for(auto& entry : boost::make_iterator_range(dir_it, {})) {
+            if (boost::filesystem::is_directory(entry)) {
+               children.emplace_back(entry);
+               continue;
+            }
+
+            // TODO: this should probably handle any supported format.
+            //   check to see if there is a list of file formats
+            //   like this
```

Comment:
This is my last thing, that I was not able to find a direct answer for. Is there a way (and should I do this), of checking all parsable formats to export to `.pt2` (ie, gp5 etc)

Reply:
I think the closest would be `FileFormatManager::findFormat()` - you could extend that or provide a similar function that only checks the available importers. `FileFormatManager::importFileFilter()` is similar but is geared towards building the filter for the file dialog.